sie_get/sie_get.1
.TH sie_get 1
.SH NAME
sie_get \- Example program to download data from SIE channels via the SIE-Batch API
.SH SYNOPSIS
.SS Command Invocation Options
Two command invocations options are available:

\(em
.B sie_get
.I channelNumber startDatetime endDatetime
.br
\(em
.B sie_get
.I channelNumber
.B now
.I minutesBack
.SS Arguments
.I channelNumber
should be a raw channel number, eg., 
.B 212

.I datetime values
(e.g.,
.I startDatetime 
and 
.I endDatetime 
) should be 19-character-long (YYYY-MM-DD hh:mm:ss) UTC datetime stamps 
enclosed in double quote marks.  For example,  
.B \(dq2020-01-14 10:22:14\(dq
.SH DESCRIPTION
.SS Security Information Exchange
The Security Information Exchange (SIE) is a Farsight Security, Inc. service
used to share passive DNS and other cybersecurity data with subcribing
participants. SIE is like having access to a pre-built global cybersecurity 
sensor network: you get the cybersecurity visibility you want without
the hassles of having to actually run your own network of sensor nodes.
.SS Channels
The real-time data streams available at SIE are organized into numbered 
"channels." The \(dqSIE Channel Guide\(dq lists publicly-disclosed channels 
available to approved subscribers, see:
.P
\(em
.UR https://www.domaintools.com/resources/user-guides/security-information-exchange-feed-guide/
https://www.domaintools.com/resources/user-guides/security-information-exchange-feed-guide/
.UE
.SS Traditional \(dqReal-Time\(dq SIE Access Options
SIE users have historically received broadcast channel traffic in real time 
either via:

\(em 
.B a leased blade
colocated at one of the SIE datacenters, or 

\(em 
remotely, via an 
.B encrypted tunnel.

While these methods work, it means that if you want to capture an 
hour's-worth of traffic from a channel, you either:

\(em
.B need to be routinely collecting and saving traffic for that channel, or

\(em
.B need to start collecting, wait while that data is collected from that
point forward for the specified duration, and then proceed. 

Farsight wanted an alternative for the impatient that would allow data 
scientists and other users to immediately get the SIE data they wanted 
with virtually no delay.  SIE-Batch, as described in the next subsection, 
is the solution to that problem.
.SS SIE-Batch
SIE-Batch caches recently-seen data from select SIE channels, 
thereby allowing users to easily immediately download chunks of data 
from those channels.

SIE-Batch can be used in two different ways:

\(em Via a 
.B point-and-click web page, see
.UR https://siebatch.it.fsi.io/ 
https://siebatch.it.fsi.io/ 
.UE

The SIE-Batch web site requires an SIE-Batch API key. Please contact
Farsight Security Sales at sales@farsightsecurity.com or +1-650-489-7919
for information about getting an SIE-Batch API key.

\(em In addition to the web interface, 
.B SIE-Batch also offers an API for use by developers 
interested in integrating SIE-Batch into their own 
applications (it requires users to have the same SIE-Batch API key as 
the web interface). 

The API is documented at

.UR https://siebatch.it.fsi.io/apidoc/
https://siebatch.it.fsi.io/apidoc/
.UE

We used that API to create a demo SIE-Batch CLI (command line interface) 
client \(em that's the subject of this document.

While the SIE-Batch web page is extremely easy to use, as you'd expect, the 
.B sie_get 
command line client can be very convenient if you like to work at the Un*x 
dollar sign prompt, or you want a command that you can easily invoke 
from cron.
.SH FILES
.SS SIE-Batch API key file
~/.sie-get-key.txt 
\(em Holds the user's SIE-Batch API key for 
authentication and access control purposes. Ensure this file is NOT
world or group readable (anyone with access to this API key
will have access to your SIE-Batch account): $ chmod go-rwx ~/.sie-get-key.txt
.SS Output
Output gets written to the current default directory in files such as

.nf
\(em sie-ch212-{2020-01-09@19:46:42}-{2020-01-09@20:01:42}.jsonl
\(em sie-ch208-{2020-01-09@21:39:44}-{2020-01-09@21:44:44}.nmsg
.fi

The file type is keyed to the content of the file, either JSON Lines or NMSG.

Some operating systems may change how those files are 
.B represented
in graphical file browsers. For example, in Finder under Mac OS X 
Catalina, you'll see the colons in the filenames displayed as slashes 
(although in reality the underlying filenames are unchanged).
.SS Processing The Output Files
SIE-Batch (and applications that rely on its API, such as this one)
may emit two different kinds of files, JSON Line files and nmsg files.
(Which you'll get depends on the SIE channel you're using).

.B \(em JSON Lines output files 

JSON Lines format output files can be processed with jq:

.UR https://stedolan.github.io/jq/
https://stedolan.github.io/jq/
.UE

For example, to extract a single field of interest, e.g.:

.nf
.eo
nmsgtool -r \
 "sie-ch204-{2020-01-10@08:01:32}-{2020-01-10@08:03:32}.nmsg" \
  -V SIE -T dnsdedupe -J - | jq -r '.message.rrname'
.ec
.fi

You can also use jq with nmsgtool to extract multiple fields, 
simplifying subsetting. For example, if you wanted to drop all 
records with RRSIG or NSEC3 rrtypes, and skip all .in-addr.arpa's 
rrnames, you could use:

.nf
.eo
nmsgtool -r \
 "sie-ch204-{2020-01-10@08:01:32}-{2020-01-10@08:03:32}.nmsg" \
 -V SIE -T dnsdedupe -J - | jq -r \
 '"\(.message.rrname) \(.message.rrtype)"' | grep -v " RRSIG" | \
 grep -v " NSEC3" | grep -v ".in-addr.arpa. "
.ec
.fi

.B \(em nmsg format output files

nmsg format output files from SIE-Batch can be read using nmsgtool:

.UR https://github.com/farsightsec/nmsg 
https://github.com/farsightsec/nmsg 
.UE

For example, assume we've grabbed two minutes of Ch204, and that 
output's in:

.nf
.eo
sie-ch204-{2020-01-10@08:01:32}-{2020-01-10@08:03:32}.nmsg
.ec
.fi

To turn that nmsg data into human-readable presentation format:

.nf
.eo
nmsgtool -r \
 "sie-ch204-{2020-01-10@08:01:32}-{2020-01-10@08:03:32}.nmsg" \
 -V SIE -T dnsdedupe > somefilename.txt
.ec
.fi

Turn that nmsg data into JSON Lines format:

.nf
.eo
nmsgtool -r \
 "sie-ch204-{2020-01-10@08:01:32}-{2020-01-10@08:03:32}.nmsg" \
 -V SIE -T dnsdedupe -J - > somefilename.jsonl
.ec
.fi


sie_get/sie_get.c
/* sie_get.c */

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <curl/curl.h>

/* write_data function from https://curl.haxx.se/libcurl/c/url2file.html */
/* static size_t write_data(void *ptr, size_t size, size_t nmemb, void *stream)
{ size_t  written = fwrite(ptr, size, nmemb, (FILE *) stream);
  return written; }
*/

/* function declarations */
char * getAPIkey(void);

/* main **************************************************************** */

#define BIG 200
#define EVENBIGGER 500

int main(int argc, char **argv)
{ 

  /* declarations (alphabetically ordered by variable name) */

  char arguments[EVENBIGGER];
  int chanflag = 0;
  CURL *curl_handle;
  char currentGMTtimestring[BIG];
  char enddatetime[BIG];
  time_t epoch;
  char *filetype = NULL;
  const char *fmt = "%Y-%m-%d %H:%M:%S";
  char *fullcommand;
  struct tm *gmt;
  unsigned long i;
  size_t lengthofstring1;
  size_t lengthofstring2;
  char *myapikey = NULL;
  int mysecondsback = 0;
  static char *nowliteral = "now";
  FILE *outputfile;
  char outputfilename[BIG];
  CURLcode res;
  char startdatetime[BIG];
  struct tm *start_from_offset_val;
  char string1[BIG];
  char string2[BIG];
  time_t t;
  long http_code = 0;

  /* if we don't have the number of arguments we expect, provide precis */
  if (argc != 4) {
     printf ("\nsie_get\n\n");
     printf ("ERROR: Wrong number of command line arguments supplied.\n\n");
     printf ("  Usage:   sie_get channel \"startdatetime\" \"enddatetime\"\n");
     printf ("  Example: sie_get 212 \"2020-01-07 00:13:00\" \"2020-01-07 00:28:00\"\n");
     printf ("  Notes:   UTC datetimes must within a day or so and must be double quoted.\n");
     printf ("           Zero pad any single digit months, days, hours or minutes.\n");
     printf ("           Seconds must be part of the UTC datetimes (but are ignored.)\n");
     printf ("           Need the current UTC time? Try $ date -u\n");
     printf ("or\n\n");
     printf ("  Usage:   sie_get channel \"now\" minutesBack\n");
     printf ("  Example: sie_get 212 now 20\n"); 
     printf ("  Note:    Generally: 0 < minutesBack <= 1440 (and integer valued)\n\n");
     printf ("TERMINATED\n\n");
     exit(EXIT_FAILURE);
   }

  /* get the command line arguments */
  chanflag = atoi(argv[1]);
  char *chanflagstring = strdup(argv[1]);
  strcpy(startdatetime, argv[2]);
  strcpy(enddatetime, argv[3]);

  /* save the current UTC time */
  t = time(NULL);
  gmt = gmtime(&t);
  strftime(currentGMTtimestring, sizeof(currentGMTtimestring), fmt, gmt);

  /* get and validate the user's SIE-Batch API key */
  myapikey = getAPIkey();

  /* allow either specify absolute datetime strings, or "now" and */
  /* integer minutesBack. this block handles the 2nd case */

  if (strcmp(startdatetime,nowliteral) == 0) {
     /* enddatetime is actually minutesBack field saying how far back
        we want to go from "now", so we save it as such in a new int var */
     mysecondsback = atoi(enddatetime)*60;

     /* set the endtime to the current GMT time (as a human datetime) */
     /* (replacing the number of minutes specified by the user) */
     strcpy(enddatetime, currentGMTtimestring);
     strcpy(string2, currentGMTtimestring);

     /* set the start time to the calculated earlier human datetime, */
     /* replacing "now" */
     epoch = t - mysecondsback;
     start_from_offset_val = gmtime(&epoch);
     strftime(startdatetime, sizeof(startdatetime), fmt, start_from_offset_val);
     strcpy(string1, startdatetime);

  } else {

     /* in the explicit form, we've already startdatetime and enddatetime */
     /* we just need a copy we can munge to form the filename */
     strcpy(string1, startdatetime);
     strcpy(string2, enddatetime);

  }

  /* don't like spaces in filenames. (this is a framework for fixing other */
  /* unwanted chars, should we need to do so, too). we'll fix start and end */

  strcpy(string1, startdatetime);
  lengthofstring1=strlen(string1);
  for (i = 0; i <= lengthofstring1; i++) {
     if (string1[i] == ' ') string1[i] = '@';
  }

  strcpy(string2, enddatetime);
  lengthofstring2=strlen(string2);
  for (i = 0; i <= lengthofstring2; i++) {
       if (string2[i] == ' ') string2[i] = '@';
  }

  /* we want a file extension reflective of the file's contents. A number */
  /* of channels use native NMSG format, the rest are JSON Lines */

  if ((chanflag == 204) ||
      (chanflag == 206) ||
      (chanflag == 207) ||
      (chanflag == 208) ||
      (chanflag == 221))
    filetype = ".nmsg";
  else filetype = ".jsonl";

  /* now build the output filename by concatenating elements into a name */
  strcpy(outputfilename, "sie-ch");
  strcat(outputfilename, chanflagstring);
  strcat(outputfilename, "-{");
  strcat(outputfilename, string1);
  strcat(outputfilename, "}-{");
  strcat(outputfilename, string2);
  strcat(outputfilename, "}");
  strcat(outputfilename, filetype);

  /* initialize curl. this must be called once and only once. */
  if (curl_global_init(CURL_GLOBAL_ALL) != 0) {
    printf("curl_global_init() failed\n");
    exit(EXIT_FAILURE);
  }

  /* the POST we're going to be doing needs some arguments to work */
  /* we'll assemble those arguments now */
  strcpy(arguments, "{\"apikey\":\"");
  strcat(arguments, myapikey);
  strcat(arguments, "\",\"channel\":");
  strcat(arguments, chanflagstring);
  strcat(arguments, ",\"start_time\":\"");
  strcat(arguments, startdatetime);
  strcat(arguments, "\",\"end_time\":\"");
  strcat(arguments, enddatetime);
  strcat(arguments, "\"}"); 

  printf("about to do the curl_easy_init\n");
  curl_handle = curl_easy_init();
  printf("just did the the curl_easy_init\n");

  fullcommand="https://stage-batch.dev.fsi.io/siebatchd/v1/siebatch/chfetch";

  printf("about to do the curl_easy_setopt curlopt_url\n");
  curl_easy_setopt(curl_handle, CURLOPT_URL, fullcommand);
  printf("just did the curl_easy_setopt curlopt_url\n");

  printf("about to do the curlopt_postfields\n");
  curl_easy_setopt(curl_handle, CURLOPT_POSTFIELDS, arguments);
  printf("just did the curlopt_postfields\n");

  printf("about to do the curlopt_followlocation\n");
  curl_easy_setopt(curl_handle, CURLOPT_FOLLOWLOCATION, true);
  printf("just did the curlopt_followlocation\n");

  printf("about to do the curlopt_writefunction\n");
  /* curl_easy_setopt(curl_handle, CURLOPT_WRITEFUNCTION, write_data); */
  printf("just did the curlopt_writefunction\n");

  outputfile = fopen(outputfilename, "wb");
  if (outputfile) {
    /* write the page body to this file handle */
    printf("about to do the curlopt_writedata\n");
    curl_easy_setopt(curl_handle, CURLOPT_WRITEDATA, outputfile);
    printf("just did the curlopt_writedata\n");

    double tries = 0;
    double elapsed = 0;
    bool done = false;

    while (tries != 3 && !done) {
      /* do the actual curl command */

      printf("about to do the curl_easy_perform\n");
      res = curl_easy_perform(curl_handle);
      printf("just did the curl_easy_perform\n");

      printf("about to do the CURLINFO_EFFECTIVE_URL\n");
      curl_easy_getinfo(curl_handle, CURLINFO_EFFECTIVE_URL, &fullcommand);
      printf("just did the CURLINFO_EFFECTIVE_URL\n");

      printf("about to do the CURLINFO_RESPONSE_CODE\n");
      curl_easy_getinfo(curl_handle, CURLINFO_RESPONSE_CODE, &http_code);
      printf("just did the CURLINFO_RESPONSE_CODE\n");

      printf("about to do the CURLINFO_TOTAL_TIME\n");
      curl_easy_getinfo(curl_handle, CURLINFO_TOTAL_TIME, &elapsed);
      printf("just did the CURLINFO_TOTAL_TIME, elapsed=%f\n", elapsed);

      curl_off_t cl;
      res = curl_easy_getinfo(curl_handle, 
          CURLINFO_CONTENT_LENGTH_DOWNLOAD_T, &cl);
      if(!res) {
        printf("Download size: %" CURL_FORMAT_CURL_OFF_T "\n", cl);
      }

      if (res != CURLE_OK || http_code != 200) {
         tries++;
         printf("tries = %f\n",tries);
         }
      else
         {
         printf("completed successfully\n");
         done = true;
         }

      if (res != CURLE_OK || http_code != 200) {
         if (tries == 3) {
            /* too many tries */
            return(EXIT_FAILURE);
         } else {
           /* curl failed, retry */
           printf ("retrying... %f\n", tries);
         }
       }
    }

   fclose(outputfile);
  }
  
  curl_easy_cleanup(curl_handle);
  curl_global_cleanup();
  free(myapikey);
  return (EXIT_SUCCESS);
}
